plugins/rts54hub/fu-rts54hub-device.c

```diff
@@ -0,0 +1,419 @@
+					    0x01,		/* value */
+					    0x0,		/* idx */
+					    NULL, 0,		/* data */
+					    NULL,		/* actual */
+					    FU_RTS54HUB_DEVICE_TIMEOUT_AUTH,
+					    NULL, error)) {
+		g_prefix_error (error, "failed to authenticate: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_device_erase_flash (FuRts54HubDevice *self,
+				guint8 erase_type,
+				GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (self));
+	if (!g_usb_device_control_transfer (usb_device,
+					    G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					    G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					    G_USB_DEVICE_RECIPIENT_DEVICE,
+					    0xC0 + 0x28,	/* request */
+					    erase_type * 256,	/* value */
+					    0x0,		/* idx */
+					    NULL, 0,		/* data */
+					    NULL,		/* actual */
+					    FU_RTS54HUB_DEVICE_TIMEOUT_ERASE,
+					    NULL, error)) {
+		g_prefix_error (error, "failed to erase flash: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_device_vendor_cmd (FuRts54HubDevice *self, guint8 value, GError **error)
+{
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (self));
+
+	/* don't set something that's already set */
+	if (self->vendor_cmd == value) {
+		g_debug ("skipping vendor command 0x%02x as already set", value);
+		return TRUE;
+	}
+	if (!g_usb_device_control_transfer (usb_device,
+					    G_USB_DEVICE_DIRECTION_HOST_TO_DEVICE,
+					    G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					    G_USB_DEVICE_RECIPIENT_DEVICE,
+					    0x02,		/* request */
+					    value,		/* value */
+					    0x0bda,		/* idx */
+					    NULL, 0,		/* data */
+					    NULL,		/* actual */
+					    FU_RTS54HUB_DEVICE_TIMEOUT,
+					    NULL, error)) {
+		g_prefix_error (error, "failed to issue vendor cmd 0x%02x: ", value);
+		return FALSE;
+	}
+	self->vendor_cmd = value;
+	return TRUE;
+}
+
+static gboolean
+fu_rts54hub_device_ensure_status (FuRts54HubDevice *self, GError **error)
+{
+	guint8 data[FU_RTS54HUB_DEVICE_STATUS_LEN] = { 0 };
+	GUsbDevice *usb_device = fu_usb_device_get_dev (FU_USB_DEVICE (self));
+	gsize actual_len = 0;
+
+	if (!g_usb_device_control_transfer (usb_device,
+					    G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					    G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					    G_USB_DEVICE_RECIPIENT_DEVICE,
+					    0x09,		/* request */
+					    0x0,		/* value */
+					    0x0,		/* idx */
+					    data, sizeof(data),
+					    &actual_len,	/* actual */
+					    FU_RTS54HUB_DEVICE_TIMEOUT,
+					    NULL, error)) {
+		g_prefix_error (error, "failed to get status: ");
+		return FALSE;
+	}
+	if (actual_len != FU_RTS54HUB_DEVICE_STATUS_LEN) {
+		g_set_error (error, G_IO_ERROR, G_IO_ERROR_INVALID_DATA,
+			     "only read %" G_GSIZE_FORMAT "bytes", actual_len);
+		return FALSE;
+	}
+
+	/* check the hardware capabilities */
+	self->dual_bank = (data[7] & 0x80) == 0x80;
+	self->fw_auth = (data[13] & 0x02) > 0;
+	self->is_eeprom = (data[15] & 0x02) > 0;
```

Comment:
Is this actually has_eeprom? As in there is a SPI part attached?

Reply:
we only support dual bank update on fwupd, if Hub runs on RomCode, it is very abnormal on user side, so we don't want to upload update flow that from RomCode on fwupd.
Our fw can make sure at least one bank fw is valid, so if it happen it will a big problem maybe flash damage or ...

Reply:
Okay, that's useful, thanks. I'll update the name of the variable and set the update_error.